pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1790,9 +1796,15 @@ private void batchMessageAndSend() {
                 }
             } catch (PulsarClientException e) {
                 Thread.currentThread().interrupt();
-                semaphore.ifPresent(s -> s.release(batchMessageContainer.getNumMessagesInBatch()));
+                semaphore.ifPresent(s -> s.release(Math.min(
+                        maxPermits - s.availablePermits(),
+                        batchMessageContainer.getNumMessagesInBatch()))
```

Comment:
This is just masking the problem. If the counts are out of sync, this won't be helping either as it will not put them back in sync. We need to identify the reason for why they're out of sync.

Reply:
This limits this issue https://github.com/apache/pulsar/issues/12151 happening in prod. Since we implementing a max limit.

Reply:
I don't think it limits the issue. Simply it's masking it by making the semaphore useless, as in that the acquire/release are going to be completely either always blocking or never blocking.

Reply:
So, it's actually preferable to have an exception that identifies that we have an issue (provided that we can restart from there) rather than just masking the problem. That is, until a proper solution for the root cause is implemented.

Reply:
Let's limit the count but add a logger , that will prevent the blow up in production but still record the issue.

Reply:
I don't think limiting the counter is a good idea. Making sure we recover from the exception is a better approach.

Reply:
creating a bounded semaphore seems to be a popular approach to safeguard production that's what I tried to emulate.
https://github.com/googleapis/gax-java/blob/49efdc38f2d6ca3838dcecee63dc6cd07f6c2b23/gax/src/main/java/com/google/api/gax/batching/BlockingSemaphore.java

Reply:
The terminology is not correct here. The semaphore *is* bounded (it's its only purpose): it only allows a number of acquire.

The problem is if we're trying to release it more times than it was acquired, but that doesn't make it unbounded.

Reply:
I try and limit the total number of permits that can be held. Regardless of release requests.

It's similar to this https://github.com/googleapis/gax-java/blob/49efdc38f2/gax/src/main/java/com/google/api/gax/batching/NonBlockingSemaphore.java but I am not throwing any exceptions but silently disgarding the requests. I think logging the makes sense though.